src/library_sockfs.js

```diff
@@ -5,6 +5,156 @@
+        throw new FS.ErrnoError({{{ cDefs.EPROTONOSUPPORT }}});
+      }
+
+      var sock = {
+        family,
+        type,
+        protocol,
+        sock_ops: ENDOR_SOCKFS.sock_ops,
+      };
+
+      // create the filesystem node to store the socket structure
+      var name = ENDOR_SOCKFS.nextname();
+      var node = FS.createNode(ENDOR_SOCKFS.root, name, {{{ cDefs.S_IFSOCK }}}, 0);
+      node.sock = sock;
+
+      // and the wrapping stream that enables library functions such
+      // as read and write to indirectly interact with the socket
+      var stream = FS.createStream({
+        path: name,
+        node,
+        flags: {{{ cDefs.O_RDWR }}},
+        seekable: false,
+        stream_ops: ENDOR_SOCKFS.stream_ops
+      });
+      // map the new stream to the socket structure (sockets have a 1:1
+      // relationship with a stream)
+      sock.stream = stream;
+
+      return sock;
+    },
+    getSocket(fd) {
+      var stream = FS.getStream(fd);
+      if (!stream || !FS.isSocket(stream.node.mode)) {
+        return null;
+      }
+      return stream.node.sock;
+    },
+    stream_ops: {
+      poll(stream) {
+        var sock = stream.node.sock;
+        return sock.sock_ops.poll(sock);
+      },
+      ioctl(stream, request, varargs) {
+        var sock = stream.node.sock;
+        return sock.sock_ops.ioctl(sock, request, varargs);
+      },
+      read(stream, buffer, offset, length, position /* ignored */) {
+        var sock = stream.node.sock;
+        var msg = sock.sock_ops.recvmsg(sock, length);
+        buffer.set(msg, offset);
+        return msg.length;
+      },
+      write(stream, buffer, offset, length, position /* ignored */) {
+        var sock = stream.node.sock;
+        return sock.sock_ops.sendmsg(sock, buffer, offset, length);
+      },
+      close(stream) {
+        var sock = stream.node.sock;
+        sock.sock_ops.close(sock);
+      },
+    },
+    nextname() {
+      if (!ENDOR_SOCKFS.nextname.current) {
+        ENDOR_SOCKFS.nextname.current = 0;
+      }
+      return 'endor_socket[' + (ENDOR_SOCKFS.nextname.current++) + ']';
+    },
+    websocket_sock_ops: {
+      createPeer(sock, addr, port) {},
+      getPeer(sock, addr, port) {},
+      addPeer(sock, peer) {},
+      removePeer(sock, peer) {},
+      handlePeerEvents(sock, peer) {},
+    },
+    sock_ops: {
+      poll(sock) {
+        // No polling support
+        return 0;
+      },
+      ioctl(sock, request, arg) {
+        // No ioctl is supported
+        return {{{ cDefs.EINVAL }}};
+      },
+      close(sock) {
+        return 0;
+      },
+      bind(sock, addr, port) {
+        // Client-side only for now
+        throw new FS.ErrnoError({{{ cDefs.EOPNOTSUPP }}});
+      },
+      connect(sock, addr, port) {
+        // Our universe --for now-- is ["192.168.10.20/32"]
+        if (addr != "192.168.10.20") {
+          throw new FS.ErrnoError({{{ cDefs.EHOSTUNREACH }}});
+        }
```

Comment:
+1 for limiting this for now. However, I would be more flexible with the limitation as we might add more VMs in the future:

```suggestion
        if (!addr.startsWith("192.168.10.")) {
          throw new FS.ErrnoError({{{ cDefs.EHOSTUNREACH }}});
        }
```